### PR TITLE
Esp32 support

### PR DESCRIPTION
**Pull request by mxtommy**
Not going to merge until I test on actual hardware and confirm everything still works on both esp8266 and esp32. Would love any feedback though! 

Added an "env" in platformio for esp32. (I'm using WROVER 16MB module). 

Many library changes. Quick overview:

- Use the dev version of WifiManager. It supports both esp32 and esp8266 and is a bit more clean :)
- Use uSSDP library for discovery as there are versions for each. discovery.cpp is a bit messy though...
- ESP32 has a few differences in libraries, the lib_deps takes care to download right one

Had to sprinkle ifdefs around to take care of differences between 8266 and 32.

Compiling in ArduinoIDE for esp8266 works (after adding uSSDP library). Have't tested esp32.
Compiling in PlatformIO for both esp8266 and esp32 work.

Uploaded it to WROVER module and basic stuff works :) (connect to wifi, and systemhz working)

:)

### CODE REVIEW

**Comment by ba58smith** [comment]
@mxtommy - you wrote "ESP32 has a few differences in libraries, the lib_deps takes care to download right one." Is that a PlatformIO thing? I seem to recall PIO does a good job of checking dependencies for libraries. What do I need to do to try it w/ Arduino IDE, other than add the uSSDP library?

**Comment by mxtommy** [comment]
yes lib_deps is related to PlatformIO :) Compiling under ArduinoIDE should work fine after adding uSSDP lib. Note that the version of uSSDP is different between esp8266 and esp32. :) The one in the libraries folder of this repo is for esp8266.

If you want to download directly:
https://github.com/mxtommy/uSSDP (for esp8266, has a bugfix that hasn't been updated upstream)
https://github.com/marian-craciunescu/uSSDP-ESP32

The WifiManager Library is also using the development branch. I think it compiles with the main version as well though under esp8266. The development branch is mostly for esp32 support.
https://github.com/tzapu/WiFiManager/tree/development

**Comment by ba58smith** [comment]
Trying to get the ESP32 version to compile for the ESP8266 and then the ESP32. Here's what I did:

Downloaded the ESP32 version of SigkSens.
Renamed the folders and .ino file to SigkSensESP32.
Installed the uSSDP library from the .zip file included in the SigkSens ESP32 repo.
Selected the WEMOS D1 mini Lite with 64k SPIFFS and compiled - no warnings nor errors.
Selected the WEMOS D1 mini and R2 with 1M SPIFFS and compiled - no warning nor errors.
Installed WiFiManager-dev.zip (downloaded from GitHub).
Installed uSSDP-ESP32.zip (downloaded from GitHub).
Added the following to File _ Preferences _ Additional Boards Manager URLs:
     https://dl.espressif.com/dl/package_esp32_index.json
Installed the esp32 board from Arduino Boards Manager
Selected ESP32 Dev Module as the board, and set the Partition Scheme to "Minimal SPIFFS."
Tried to compile, got this in orange text:
```In file included from /tmp/arduino_build_262810/sketch/src/sensors/systemHz/../sensorInfo.h:8:0,
                 from /tmp/arduino_build_262810/sketch/src/sensors/systemHz/../sensorStorage.h:6,
                 from /tmp/arduino_build_262810/sketch/src/sensors/systemHz/systemHz.h:5,
Multiple libraries were found for "WiFi.h"
 Used: /home/butch/.arduino15/packages/esp32/hardware/esp32/1.0.0/libraries/WiFi
 Not used: /home/butch/arduino-1.8.5/libraries/WiFi
                 from /home/butch/Documents/Arduino/mySketches/SigkSensESP32/SigkSensESP32/SigkSensESP32.ino:33:
/home/butch/Documents/Arduino/mySketches/libraries/ESPAsyncWebServer-master/src/ESPAsyncWebServer.h:33:22: fatal error: AsyncTCP.h: No such file or directory
compilation terminated.
```

The obvious thing is the multiple libraries found for "WiFi.h", but it appears it's using the esp32 version, so maybe that's not the problem? (I have "Show compiler warnings" set to "NONE" in the Arduino IDE, so the multiple libraries message seems to be an error, not just a warning - but maybe not?)

Maybe it's the "No such file or directory" for AsyncTCP.h. That would certainly shut down the compiler. When I search for that file on my entire computer, I find the following. Notice that it's only in one of the PlatformIO folders, not in any of my Arduino folders. Do you think that's the problem?
image

**Comment by ba58smith** [comment]
Did a little digging, and it seems that the ESPAsyncWebServer library is based on the AsyncTCP library, and at some point, it includes it. So I installed the AsyncTCP library (from a .ZIP file downloaded from GitHub), and tried to compile again. This time, I get:

```In file included from /tmp/arduino_build_262810/sketch/src/net/discovery.cpp:8:0:
/home/butch/Documents/Arduino/mySketches/libraries/uSSDP-master/uSSDP.h:11:25: fatal error: ESP8266WiFi.h: No such file or directory
compilation terminated.
```
This doesn't make sense to me, because ESP8266WiFi.h is located here on my hard drive:
/home/butch/.arduino15/packages/esp8266/hardware/esp8266/2.4.2/libraries/ESP8266WiFi/src/ESP8266WiFi.h
... and that should be a location that the Arduino IDE searches when looking for #include files, I think. Any thoughts?

**Comment by mxtommy** [comment]
Ah yes, the Async library is different between ESP32/ESP8266. AsyncTCP on ESP32 vs ESPAsyncTCP on ESP8266.... 

Also note I'm using specific versions of the esp32 Core (#feature/stage branch), and AsyncTCP libraries  (idf-update branch).

I will say libraries for ESP8266 seem a bit more "mature" :D

For uSSDP, I think the IDE is being confused by having both libraries installed? It seems to be including the ESP8266 core libraries when you're compiling for ESP32

Thanks for testing :D

**Comment by ba58smith** [comment]
@mxtommy - You wrote "Also note I'm using specific versions of the esp32 Core (#feature/stage branch), and AsyncTCP libraries (idf-update branch)." I don't understand. Is this something I need to do, too? Can you elaborate? (UPDATE: Oh, now I see. But how did you know to use those branches? I deleted the AsyncTCP-master library and replaced it with the idf-update branch, but the #feature/stage branch of the esp32 Core is for PlatformIO, so I didn't do anything with that. Is that correct?)

And you wrote: "For uSSDP, I think the IDE is being confused by having both libraries installed? It seems to be including the ESP8266 core libraries when you're compiling for ESP32." Do you think I should uninstall one of the ESP8266 library(ies) before compiling for the ESP32? Is there a cleaner way than that to have SigkSens be able to work with both boards?

UPDATE: Since I'm testing for the esp32, I deleted the uSSDP library folder and added the uSSDP-ESP32 library, but the Arduino IDE wouldn't recognize it. Turns out they forgot a line in the library.properties file:
`paragraph=`
I added that, and that compiler error went away (and I added that as an Issue to the uSSDP-ESP32 GitHub page).

Now I'm getting another error, about another missing file: FunctionalInterrupt.h. But someone else got that recently, so I'll dig into that next.

**Comment by ba58smith** [comment]
The compiler found two references to `#include <FunctionalInterrupt.h>`, which is, I believe, an ESP8266-only file, so I used:
```
#ifdef ESP8266
#include <FunctionalInterrupt.h>
#endif 
```
to get around those. They are at:
digitalIn.cpp, line 11
ReactESP.cpp, line 4

Tried to compile, got `LED_BUILTIN not defined in this scope`, on config.h line 66. (So I changed the line to:
`#define LED_PIN 13`
just to get around that error. I don't know what the real fix is for this, though, so it still needs to be addressed.

Tried to compile, and got this, which is way above my pay grade, so I'm punting to @mxtommy and @mairas:
/tmp/arduino_build_214746/sketch/src/sensors/digitalIn/digitalIn.cpp: In function 'void setupDigitalIn(bool&)':
/tmp/arduino_build_214746/sketch/src/sensors/digitalIn/digitalIn.cpp:121:96: error: cannot convert 'std::_Bind_helper<false, void (&)(int), unsigned int&>::type {aka std::_Bind<void (*(unsigned int))(int)>}' to 'void (*)()' for argument '2' to 'void attachInterrupt(uint8_t, void (*)(), int)'
     attachInterrupt(digitalPins[index], std::bind(DigitalPinISR, index), digitalPinModes[index]);

**Comment by mxtommy** [comment]
Ok for Arduino IDE, it's the development release.  

So the boards link is https://dl.espressif.com/dl/package_esp32_dev_index.json instead of https://dl.espressif.com/dl/package_esp32_index.json

That should handle all the stuff regarding FunctionalInterrupt (It was added since last stable release) and also those last compile errors you saw (as they're using the FunctionInterrupt stuff)

The LED_BUILTIN define is something that is defined by the board you select in the IDE, ESP32's don't define it so you need to set it as you did :)

**Comment by ba58smith** [comment]
Progress. I used the dev version of the boards manager url, and selected the ESP32 WROVER MODULE, and I un-did my two changes to get around the FunctionalInterrupt errors, and tried again. This time, I get the following, and again, I have to punt to you guys:
```
In file included from /tmp/arduino_build_247802/sketch/src/sensors/oneWire/oneWire.cpp:9:0:
/home/butch/Documents/Arduino/mySketches/libraries/MAX31850_OneWire/OneWire.h:130:2: error: #error "Please define I/O register types here"
 #error "Please define I/O register types here"
  ^
/home/butch/Documents/Arduino/mySketches/libraries/MAX31850_OneWire/OneWire.h:137:5: error: 'IO_REG_TYPE' does not name a type
     IO_REG_TYPE bitmask;
     ^
/home/butch/Documents/Arduino/mySketches/libraries/MAX31850_OneWire/OneWire.h:138:14: error: 'IO_REG_TYPE' does not name a type
     volatile IO_REG_TYPE *baseReg;
```

**Comment by mxtommy** [comment]
D'oh, the MAX31850_OneWire library doesn't support esp32. We can use the default OneWire library though. 

https://github.com/PaulStoffregen/OneWire

That's what platformIO is using actually now that I look at it, and compiles fine. Disadvantage from what I understand it doesn't support the MAX31850, (small board that lets you use a K-type thermocouple for high temps). I had planned to use that board to measure exhaust gas temps on motor. Not sure what to do for this :) Maybe I can do a pull request on the official OneWire repo to get support for MAX31850. Will check later at home :)

**Comment by ba58smith** [comment]
IT COMPILED! (After installing the OneWire library with the Library Manager, and then dealing with LED_BUILTIN on two lines in configReset.cpp - changed them to LED_PIN.)

Now I'll try to actually load it onto one of my ESP32 boards. One of them I have isn't listed, but at least one other one is, so I'll try them all.

**Comment by ba58smith** [comment]
Can't upload to any of my ESP32 boards - keep getting "'Serial' object has no attribute 'dtr'. Seems the fix is to install version 3 or later of esptool.py (or maybe it's python-serial?) on my computer. Before I do that - anything you guys know about doing that that I need to be aware / afraid of? Can you tell me what version you have installed on your computer?

**Comment by mxtommy** [comment]
Great that it compiles :D 

Going off of https://github.com/espressif/esptool/issues/324 it's PySerial that needs to be version 3. They list some possible solutions there. :) What are you compiling on? :)

**Comment by ba58smith** [comment]
I got tired of waiting (HA!), so I went ahead and updated python-serial and esptool.py as per the instructions on this page: https://github.com/espressif/esptool/issues/324

I am running Linux Mint 17 on my computer.

Then I could upload and run SigkSens! I did the configuration, but it didn't save - I had chosen the "Minimall SPIFFS" option, but apparently it didn't give me enough SPIFFS space. So I changed to "Default" (for the "Partition scheme", for this particular board), tried it all again, and this time, it worked. After getting and setting the security token, I am now seeing live data in the Signal K Instrument Panel!

But I have no sensors connected. I can hook up a OneWire temp sensor fairly easily, so I'll try that. But with no actual sensors, the Sketch uses 1,165,078 bytes - that's way bigger than the ESP8266 Sketch - more than twice as big. Is that what you see, too?

**Comment by ba58smith** [comment]
@mxtommy - What options did you choose for your ESP32 Wrover Module for "Flash mode" and "Partition scheme"?

**Comment by ba58smith** [comment]
SigkSens for ESP32 - Summary of Differences (from ESP8266) and Known Issues

1. Need to install a uSSDP library. The trick is, it needs to be the correct one. There's one for the ESP8266 (uSSDP), and one for the ESP32 (uSSDP-ESP32). To be safe, have only the correct one (you can tell by the folder name) in your /sketchbook/libraries folder, depending on what board you're compiling for.

2. Need to install the dev version of WiFiManager (WiFiManager-dev). It seems to work for ESP8266 and ESP32. https://github.com/tzapu/WiFiManager/tree/development

3. Need to install the AsyncTCP library, but a very specific branch (idf-update branch). https://github.com/me-no-dev/AsyncTCP/tree/idf-update

4. Need to use a dev version of the ESP32 board, which you get by using this url in the Additional Boards Manager URL's in the Arduino IDE:
https://dl.espressif.com/dl/package_esp32_dev_index.json

5. SigkSens has (I think) three references to LED_BUILTIN, which isn't defined for the ESP32. Need to deal with that. (config.h, and 2 in configReset.cpp) (Fixed with PR https://github.com/mxtommy/SigkSens/pull/69)

6. The MAX31850 OneWire library doesn't support ESP32, but the OneWire library does. (UPDATE: see Issue https://github.com/mxtommy/SigkSens/issues/67.) 

7. Need to update the version of esptool.py and python-serial on the computer you're doing development on, in order to communicate with the ESP32 (to download the Sketch to the board).

**Comment by ba58smith** [comment]
I have a Heltec WIFI_Kit_32 board with a built-in OLED screen (.96" diagonal - pretty cool little screen!). I selected it from the Boards list, compiled, and downloaded, no problem. Configured the network, rebooted it, got a security token and loaded it, and boom! Data is flowing to the SignalK Instrument Panel! One oddity - the hostname (the name I gave my board when I did the configuration - HeltecOLED) - didn't save, even though the SSID and pw for my network did save. I used /setNewHostname to set it, and now it appears in my serial monitor's output on reboot, but still doesn't show up in the name of the sensor in the Instrument Panel. Instead of sensors.HeltecOLED.freeMem, it displays only sensors..freeMem. Any ideas why?

**Comment by mxtommy** [comment]
I'll have a look at the code that set's the hostname initially :) It likely failed. The paths for systemhz are set/saved on first bootup, (so empty). Using setNewHostname to set it after doesn't update the paths :)

File: platformio.ini [hunk 1]
```diff
@@ -27,7 +27,7 @@ lib_deps =
     ESP8266mDNS
     me-no-dev/uSSDP
     DNSServer
-    WifiManager
+    https://github.com/tzapu/WiFiManager#bbe984c93b71601aad9f50f765ba9a255478bff3
```
Comment (mairas) [inline platformio.ini:30]: Even if the ESP32 version would require a specific branch, I wouldn't pin library versions to particular commits, especially for ESP8266.

File: platformio.ini [hunk 2]
```diff
@@ -40,26 +40,27 @@ lib_deps =
...
 framework = arduino
 board = lolin32
 monitor_speed = 115200
 board_build.extra_flags = -DARDUINO_ESP32_DEV -DBOARD_HAS_PSRAM -mfix-esp32-psram-cache-issue
 board_upload.flash_size = 16MB
 board_upload.maximum_size = 16777216
+board_build.partitions = default_16MB.csv
 lib_deps =
     ReactESP
     https://github.com/tzapu/WiFiManager#bbe984c93b71601aad9f50f765ba9a255478bff3
```
Comment (mairas) [inline platformio.ini:53]: Instead of pinning to one commit, I'd pin to the branch providing ESP32 support (possibly `#development`).

File: platformio.ini [hunk 3]
```diff
@@ -40,26 +40,27 @@ lib_deps =
...
     ESPmDNS
     marian-craciunescu/uSSDP-ESP32
     ArduinoJson
     WebSockets
-    ESPAsyncTCP
-    ESP Async WebServer
+    me-no-dev/AsyncTCP#6fab7724abe9fb1669a77e8a346d12ec148ff19a
```
Comment (mairas) [inline platformio.ini]: Here, too: if AsyncTCP requires some branch to work, pin to that branch, not to a single commit.

File: SigkSens/FSConfig.cpp [hunk 4]
```diff
@@ -27,10 +32,17 @@ Config Save/Load/Reset
 ------------------------------------------------------------------------------
 ----------------------------------------------------------------------------*/
 
+
```
Comment (mairas) [inline SigkSens/FSConfig.cpp:35]: Nitpicking on the format: unnecessary whitespace :-)

**Comment by mairas** [comment]
I'm running a bit out of time here - didn't test this PR on any actual device yet. Overall, it looks straightforward and the amount if `#ifdef`s isn't too bad. If you can be reasonably sure that there are no regressions in ESP8266 support, I'd recommend merging this in and then fixing any new bugs with smaller PRs. Maintaining a large PR separately is a major pain in the butt. :-D

**Comment by ba58smith** [comment]
I think I know how to add ESP32 support to the MAX31850_OneWire library. I'm going to put the details here, so I don't lose them. UPDATE: See https://github.com/mxtommy/SigkSens/issues/67

File: platformio.ini (hunk 3 above)
Comment (mxtommy) [reply platformio.ini]: Changed to Branch as should be ok :)

File: platformio.ini (hunk 1 above)
Comment (mxtommy) [reply platformio.ini:30]: For this one, the latest Development branch commit doesn't even pass their own CI checks :( I'd feel better leaving it at this commit for now. I'll revisit in the future when hopefully they release esp32 support, then we can put the good branch/master :)
